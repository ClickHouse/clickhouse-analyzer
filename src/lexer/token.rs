use std::fmt;

/// ClickHouse Tokens, same as the original
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenKind {
    // Base tokens
    Whitespace,
    Comment,

    BareWord,       // Keywords or identifiers
    Number,         // Numeric literals
    StringLiteral,  // String literals with single quotes
    QuotedIdentifier, // Double-quoted or backtick-quoted identifiers

    // Brackets
    OpeningRoundBracket,
    ClosingRoundBracket,
    OpeningSquareBracket,
    ClosingSquareBracket,
    OpeningCurlyBrace,
    ClosingCurlyBrace,

    // Punctuation
    Comma,
    Semicolon,
    VerticalDelimiter, // \G
    Dot,

    // Operators and special symbols
    Asterisk,
    HereDoc,
    DollarSign,
    Plus,
    Minus,
    Slash,
    Percent,
    Arrow,
    QuestionMark,
    Colon,
    Caret,
    DoubleColon,
    Equals,
    NotEquals,
    Less,
    Greater,
    LessOrEquals,
    GreaterOrEquals,
    Spaceship,     // <=>
    PipeMark,
    Concatenation, // ||

    // MySQL-style variables
    At,
    DoubleAt,

    // End of stream
    EndOfStream,

    // Error tokens
    Error,
    ErrorMultilineCommentIsNotClosed,
    ErrorSingleQuoteIsNotClosed,
    ErrorDoubleQuoteIsNotClosed,
    ErrorBackQuoteIsNotClosed,
    ErrorSingleExclamationMark,
    ErrorSinglePipeMark,
    ErrorWrongNumber,
    ErrorMaxQuerySizeExceeded,
    
    // Temporary hack for WHERE operators
    And, // AND
    Or // OR
}

impl fmt::Display for TokenKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            TokenKind::BareWord => write!(f, "identifier or keyword"),
            TokenKind::Number => write!(f, "number"),
            TokenKind::StringLiteral => write!(f, "string literal"),
            TokenKind::QuotedIdentifier => write!(f, "quoted identifier"),
            TokenKind::OpeningRoundBracket => write!(f, "("),
            TokenKind::ClosingRoundBracket => write!(f, ")"),
            // All other token types
            _ => write!(f, "{:?}", self),
        }
    }
}

/// Structure representing a token in the SQL
#[derive(Debug, Clone)]
pub struct Token {
    pub kind: TokenKind,
    pub text: String,
    pub start: usize,  // Start position in the source
    pub end: usize,    // End position in the source
    pub line: usize,   // Line number
    pub column: usize, // Column number
}

impl Token {
    pub fn new(kind: TokenKind, value: String, start: usize, end: usize, line: usize, column: usize) -> Self {
        Self {
            kind,
            text: value,
            start,
            end,
            line,
            column,
        }
    }
}
