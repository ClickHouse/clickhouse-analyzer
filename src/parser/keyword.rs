#[derive(Debug, Copy, Clone)]
pub enum Keyword {
    With,
    Select,
    From,
    Order,
    By,
    As,
    Where,
    And,
    Or,
    Limit,
}

impl Keyword {
    pub fn as_str(&self) -> &'static str {
        match self {
            Keyword::With => "WITH",
            Keyword::Select => "SELECT",
            Keyword::From => "FROM",
            Keyword::Order => "ORDER",
            Keyword::By => "BY",
            Keyword::As => "AS",
            Keyword::Where => "WHERE",
            Keyword::And => "AND",
            Keyword::Or => "OR",
            Keyword::Limit => "LIMIT",
        }
    }
}
