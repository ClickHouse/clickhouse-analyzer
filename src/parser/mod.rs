pub mod keyword;
pub mod parser;
mod parsers;
pub mod tree;
