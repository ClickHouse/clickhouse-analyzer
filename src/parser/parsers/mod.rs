pub mod column_type;
pub mod expression;
pub mod select;
