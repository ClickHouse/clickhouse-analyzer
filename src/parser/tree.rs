use std::fmt;
use crate::lexer::token::{Token, TokenKind};

#[derive(Debug, PartialEq, Copy, Clone)]
pub enum TreeKind {
    // Error handling
    ErrorTree,

    // Root and container nodes
    File,      // Root of the tree
    QueryList, // Multiple queries separated by semicolons

    // Statements
    SelectStatement,
    InsertStatement,
    UpdateStatement,
    DeleteStatement,
    CreateStatement,
    AlterStatement,
    DropStatement,
    TruncateStatement,
    RenameStatement,
    ShowStatement,
    UseStatement,
    SetStatement,
    OptimizeStatement,
    SystemStatement,

    // SELECT statement components
    WithClause,      // WITH subqueries
    SelectClause,    // SELECT columns
    FromClause,      // FROM tables
    JoinClause,      // Any join
    ArrayJoinClause, // ARRAY JOIN
    PrewhereClause,  // PREWHERE (ClickHouse specific)
    WhereClause,     // WHERE conditions
    GroupByClause,   // GROUP BY columns
    HavingClause,    // HAVING conditions
    OrderByClause,   // ORDER BY columns
    LimitByClause,   // LIMIT BY (ClickHouse specific)
    LimitClause,     // LIMIT rows
    SettingsClause,  // SETTINGS (ClickHouse specific)

    // CREATE statement components
    TableDefinition,
    DatabaseDefinition,
    ViewDefinition,
    MaterializedViewDefinition,
    DictionaryDefinition,

    // Column definition components
    ColumnDefinition,
    ColumnTypeDefinition,
    ColumnConstraint,
    TableConstraint,

    // Table components
    TableIdentifier, // Database.Table
    TableExpression, // Table or subquery
    TableFunction,   // Table function like merge()

    // Expressions
    Expression,         // Generic expression
    Asterisk,           // Asterisk expression
    ColumnReference,    // Column reference
    ColumnAlias,        // Column alias
    QualifiedName,      // Database.table.column
    FunctionCall,       // Function call
    AggregateFunction,  // Aggregate function with possible DISTINCT/ALL
    CastExpression,     // CAST(x AS type)
    CaseExpression,     // CASE WHEN ... THEN
    BinaryExpression,   // a + b, a AND b, etc.
    UnaryExpression,    // NOT a, -b
    BetweenExpression,  // a BETWEEN b AND c
    InExpression,       // a IN (b, c)
    TupleExpression,    // (a, b, c)
    ArrayExpression,    // [a, b, c]
    MapExpression,      // {a:b, c:d}
    SubqueryExpression, // (SELECT ...)
    LambdaExpression,   // x -> expr

    // Literals
    NumberLiteral,  // 123, 3.14
    StringLiteral,  // 'text'
    DateLiteral,    // DATE '2023-01-01'
    BooleanLiteral, // true, false
    NullLiteral,    // NULL

    // Lists and collections
    ColumnList,     // List of columns
    ExpressionList, // List of expressions
    OrderByList,    // List of ORDER BY items
    GroupByList,    // List of GROUP BY items
    SettingList,    // List of settings

    // Join specific
    JoinType,       // LEFT, INNER, etc.
    JoinConstraint, // ON clause or USING clause

    // Items that are part of larger constructs
    OrderByItem,        // column ASC NULLS FIRST
    WithExpressionItem, // name AS (subquery)

    // Data type definitions
    DataType, // Int32, String, etc.
    DataTypeParameters,
    NestedDataType, // Array(Int32), Tuple(...)
    EnumValue,      // 'value' = 1

    // ClickHouse specific
    PartitionExpression, // PARTITION BY expr
    SampleExpression,    // SAMPLE n

    // Trivia
    Whitespace,
    LineComment,  // -- comment
    BlockComment, // /* comment */
}

pub struct Tree {
    pub kind: TreeKind,
    pub children: Vec<Child>,
}

pub enum Child {
    Token(Token),
    Tree(Tree),
}

impl Child {
    pub fn is_token(&self) -> bool {
        matches!(self, Child::Token(_))
    }
    
    pub fn is_tree(&self) -> bool {
        matches!(self, Child::Token(_))
    }

    pub fn get_token_with_kind(&self, kind: TokenKind) -> Option<&Token> {
        match self {
            Child::Token(token) if token.kind == kind => Some(token),
            _ => None,
        }
    }
    
    pub fn get_tree_with_kind(&self, kind: TreeKind) -> Option<&Tree> {
        match self {
            Child::Tree(tree) if tree.kind == kind => Some(tree),
            _ => None,
        }
    }
}

pub trait ChildOptionExt {
    fn get_token_with_kind(&self, kind: TokenKind) -> Option<&Token>;
    fn get_tree_with_kind(&self, kind: TreeKind) -> Option<&Tree>;
}

impl ChildOptionExt for Option<&Child> {
    fn get_token_with_kind(&self, kind: TokenKind) -> Option<&Token> {
        match self {
            Some(Child::Token(token)) if token.kind == kind => Some(token),
            _ => None,
        }
    }
    
    fn get_tree_with_kind(&self, kind: TreeKind) -> Option<&Tree> {
        match self {
            Some(Child::Tree(tree)) if tree.kind == kind => Some(tree),
            _ => None
        }
    }
}

#[macro_export]
macro_rules! format_to {
    ($buf:expr) => ();
    ($buf:expr, $lit:literal $($arg:tt)*) => {
        { use ::std::fmt::Write as _; let _ = ::std::write!($buf, $lit $($arg)*); }
    };
}

impl Tree {
    pub fn print(&self, buf: &mut String, level: usize) {
        let indent = "  ".repeat(level);
        format_to!(buf, "{indent}{:?}\n", self.kind);
        for child in &self.children {
            match child {
                Child::Token(token) => {
                    if token.kind == TokenKind::Whitespace {
                        continue;
                    }
                    format_to!(buf, "{indent}  '{}'\n", token.text)
                }
                Child::Tree(tree) => tree.print(buf, level + 1),
            }
        }
        assert!(buf.ends_with('\n'));
    }
}

impl fmt::Debug for Tree {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut buf = String::new();
        self.print(&mut buf, 0);
        write!(f, "{}", buf)
    }
}
