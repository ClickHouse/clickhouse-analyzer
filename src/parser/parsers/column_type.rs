use crate::lexer::token::TokenKind;
use crate::parser::parser::Parser;
use crate::parser::tree::TreeKind;

pub fn parse_column_type(p: &mut Parser) {
    let m = p.open();

    if p.at(TokenKind::BareWord) {
        p.advance();
    } else {
        p.advance_with_error("Expected type for cast operator");
    }

    if p.at(TokenKind::OpeningRoundBracket) {
        let m = p.open();
        p.expect(TokenKind::OpeningRoundBracket);
        parse_column_type(p);

        while p.at(TokenKind::Comma) && !p.eof() {
            p.expect(TokenKind::Comma);
            parse_column_type(p);
        }

        p.expect(TokenKind::ClosingRoundBracket);
        p.close(m, TreeKind::DataTypeParameters);
    }

    p.close(m, TreeKind::DataType);
}
