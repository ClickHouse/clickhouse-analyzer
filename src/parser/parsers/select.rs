use crate::lexer::token::TokenKind;
use crate::parser::keyword::Keyword;
use crate::parser::parser::Parser;
use crate::parser::parsers::expression::parse_expression;
use crate::parser::tree::TreeKind;

pub fn parse_select_statement(p: &mut Parser) {
    let m = p.open();

    if p.at_keyword(Keyword::With) {
        parse_with_clause(p);
    }

    let mut parsed_early_from = false;
    if p.at_keyword(Keyword::From) {
        parse_from_clause(p);
        parsed_early_from = true;
    }

    parse_select_clause(p);

    if p.at_keyword(Keyword::From) {
        parse_from_clause(p);

        if parsed_early_from {
            p.recover_with_error("Duplicate FROM clause");
        }
    }

    if p.at_keyword(Keyword::Where) {
        let m = p.open();
        p.expect_keyword(Keyword::Where);
        parse_expression(p);
        p.close(m, TreeKind::WhereClause);
    }

    if p.at_keyword(Keyword::Order) {
        let m = p.open();
        p.expect_keyword(Keyword::Order);
        p.expect_keyword(Keyword::By);
        let m2 = p.open();
        parse_expression(p);
        p.close(m2, TreeKind::OrderByItem);
        p.close(m, TreeKind::OrderByClause);
    }

    if p.at_keyword(Keyword::Limit) {
        let m = p.open();
        p.expect_keyword(Keyword::Limit);
        parse_expression(p);
        p.close(m, TreeKind::LimitClause);
    }

    p.close(m, TreeKind::SelectStatement);
}

// Finds the end of a WITH or a SELECT
pub fn at_end_of_column_list(p: &mut Parser) -> bool {
    p.at_keyword(Keyword::Select)
        || p.at_keyword(Keyword::From)
        || p.at_keyword(Keyword::Where)
        || p.at_keyword(Keyword::Order)
        || p.at_keyword(Keyword::Limit)
}

pub fn at_select_statement(p: &mut Parser) -> bool {
    p.at_keyword(Keyword::With) || p.at_keyword(Keyword::Select) || p.at_keyword(Keyword::From)
}

pub fn parse_with_clause(p: &mut Parser) {
    let m = p.open();

    p.expect_keyword(Keyword::With);

    // Parse column list
    parse_column_list(p);

    p.close(m, TreeKind::WithClause);
}

pub fn parse_select_clause(p: &mut Parser) {
    let m = p.open();

    p.expect_keyword(Keyword::Select);

    // Parse column list
    parse_column_list(p);

    p.close(m, TreeKind::SelectClause);
}

// Parse a comma-separated list of column expressions
pub fn parse_column_list(p: &mut Parser) {
    let m = p.open();

    let mut first = true;
    while !at_end_of_column_list(p) && !p.end_of_statement() {
        if !first {
            p.expect(TokenKind::Comma);
        }
        first = false;

        parse_expression(p);

        if p.at_keyword(Keyword::As)
            || (!at_end_of_column_list(p) && p.at(TokenKind::BareWord))
            || p.at(TokenKind::QuotedIdentifier)
        {
            let m = p.open();
            if p.at_keyword(Keyword::As) {
                p.expect_keyword(Keyword::As);
            }

            if !at_end_of_column_list(p) {
                p.advance()
            } else {
                p.recover_with_error("Expected column alias");
            }

            p.close(m, TreeKind::ColumnAlias);
        }
    }

    p.close(m, TreeKind::ColumnList);
}

// Parse the FROM clause
fn parse_from_clause(p: &mut Parser) {
    let m = p.open();

    p.expect_keyword(Keyword::From);

    parse_table_reference(p);

    p.close(m, TreeKind::FromClause);
}

// Parse a table reference
fn parse_table_reference(p: &mut Parser) {
    let m = p.open();

    if p.at_any(&[TokenKind::BareWord, TokenKind::QuotedIdentifier]) {
        // Simple table name
        p.advance();

        // Handle optional database.table notation
        if p.at(TokenKind::Dot) {
            p.advance(); // Consume dot

            if p.at_any(&[TokenKind::BareWord, TokenKind::QuotedIdentifier]) {
                p.advance();
            } else {
                p.advance_with_error("Expected table name after dot");
            }
        }
    } else {
        p.advance_with_error("Expected table reference");
    }

    p.close(m, TreeKind::TableIdentifier);
}
