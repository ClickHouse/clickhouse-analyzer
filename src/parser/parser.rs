use crate::lexer::token::{Token, TokenKind};
use crate::lexer::tokenizer::tokenize_with_whitespace;
use crate::parser::keyword::Keyword;
use crate::parser::parsers::select::{at_select_statement, parse_select_statement};
use crate::parser::tree::{Child, Tree, TreeKind};
use std::cell::Cell;

#[derive(Debug)]
pub enum Event {
    Open { kind: TreeKind },
    Close,
    Advance,
}

pub struct MarkOpened {
    index: usize,
}

pub struct MarkClosed {
    index: usize,
}

pub struct Parser {
    tokens: Vec<Token>,
    pos: usize,
    fuel: Cell<u32>,
    events: Vec<Event>,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Parser {
        Parser {
            tokens,
            pos: 0,
            fuel: Cell::new(256),
            events: Vec::new(),
        }
    }

    pub fn build_tree(self) -> Tree {
        let mut tokens = self.tokens.into_iter();
        let mut events = self.events;

        assert!(matches!(events.pop(), Some(Event::Close)));
        let mut stack = Vec::new();
        for event in events {
            match event {
                Event::Open { kind } => stack.push(Tree {
                    kind,
                    children: Vec::new(),
                }),
                Event::Close => {
                    let tree = stack.pop().unwrap();
                    stack.last_mut().unwrap().children.push(Child::Tree(tree));
                }
                Event::Advance => {
                    let token = tokens.next().unwrap();
                    stack.last_mut().unwrap().children.push(Child::Token(token));
                }
            }
        }

        let tree = stack.pop().unwrap();
        assert!(stack.is_empty());
        assert!(tokens.next().is_none());
        tree
    }

    pub fn open(&mut self) -> MarkOpened {
        let mark = MarkOpened {
            index: self.events.len(),
        };
        self.events.push(Event::Open {
            kind: TreeKind::ErrorTree,
        });
        mark
    }

    pub fn open_before(&mut self, m: MarkClosed) -> MarkOpened {
        let mark = MarkOpened { index: m.index };
        self.events.insert(
            m.index,
            Event::Open {
                kind: TreeKind::ErrorTree,
            },
        );
        mark
    }

    pub fn close(&mut self, m: MarkOpened, kind: TreeKind) -> MarkClosed {
        self.events[m.index] = Event::Open { kind };
        self.events.push(Event::Close);
        MarkClosed { index: m.index }
    }

    pub fn skip_trivia(&mut self) {
        while self.at_any_with_trivia(&[TokenKind::Whitespace, TokenKind::Comment]) && !self.eof() {
            self.advance();
        }
    }

    pub fn advance(&mut self) {
        assert!(!self.eof());
        self.fuel.set(256);
        self.events.push(Event::Advance);
        println!("{}", self.nth_text_with_trivia(0));
        self.pos += 1;
    }

    pub fn recover_with_error(&mut self, error: &str) {
        let m = self.open();
        // TODO: Error reporting.
        eprintln!("{error}");
        self.close(m, TreeKind::ErrorTree);
    }

    pub fn advance_with_error(&mut self, error: &str) {
        let m = self.open();
        // TODO: Error reporting.
        eprintln!("{error}");
        if !self.eof() {
            self.advance();
        }
        self.close(m, TreeKind::ErrorTree);
    }

    pub fn eof(&self) -> bool {
        self.pos == self.tokens.len()
    }

    // Handles semicolon and eof
    pub fn end_of_statement(&mut self) -> bool {
        self.at(TokenKind::ClosingRoundBracket) || self.at(TokenKind::Semicolon) || self.eof()
    }

    pub fn nth(&mut self, lookahead: usize) -> TokenKind {
        self.skip_trivia();
        if self.fuel.get() == 0 {
            panic!("parser is stuck")
        }
        self.fuel.set(self.fuel.get() - 1);
        self.tokens
            .get(self.pos + lookahead)
            .map_or(TokenKind::EndOfStream, |it| it.kind)
    }

    pub fn nth_with_trivia(&self, lookahead: usize) -> TokenKind {
        if self.fuel.get() == 0 {
            panic!("parser is stuck")
        }
        self.fuel.set(self.fuel.get() - 1);
        self.tokens
            .get(self.pos + lookahead)
            .map_or(TokenKind::EndOfStream, |it| it.kind)
    }

    pub fn at(&mut self, kind: TokenKind) -> bool {
        self.nth(0) == kind
    }

    pub fn at_with_trivia(&mut self, kind: TokenKind) -> bool {
        self.nth_with_trivia(0) == kind
    }

    pub fn at_any(&mut self, kinds: &[TokenKind]) -> bool {
        kinds.contains(&self.nth(0))
    }

    pub fn at_any_with_trivia(&mut self, kinds: &[TokenKind]) -> bool {
        kinds.contains(&self.nth_with_trivia(0))
    }

    pub fn eat(&mut self, kind: TokenKind) -> bool {
        if self.at(kind) {
            self.advance();
            true
        } else {
            false
        }
    }

    pub fn expect(&mut self, kind: TokenKind) {
        if self.eat(kind) {
            return;
        }
        // TODO: Error reporting.
        eprintln!("expected {kind:?}");
    }

    pub fn nth_text(&mut self, lookahead: usize) -> &str {
        self.skip_trivia();
        if self.fuel.get() == 0 {
            panic!("parser is stuck")
        }
        self.fuel.set(self.fuel.get() - 1);
        self.tokens
            .get(self.pos + lookahead)
            .map_or("", |it| it.text.as_str())
    }

    pub fn nth_text_with_trivia(&mut self, lookahead: usize) -> &str {
        if self.fuel.get() == 0 {
            panic!("parser is stuck")
        }
        self.fuel.set(self.fuel.get() - 1);
        self.tokens
            .get(self.pos + lookahead)
            .map_or("", |it| it.text.as_str())
    }

    pub fn at_keyword(&mut self, keyword: Keyword) -> bool {
        self.nth(0) == TokenKind::BareWord
            && self.nth_text(0).eq_ignore_ascii_case(keyword.as_str())
    }

    pub fn eat_keyword(&mut self, keyword: Keyword) -> bool {
        if self.at_keyword(keyword) {
            self.advance();
            true
        } else {
            false
        }
    }

    pub fn expect_keyword(&mut self, keyword: Keyword) {
        if self.eat_keyword(keyword) {
            return;
        }
        // TODO: Error reporting.
        eprintln!("expected {keyword:?}");
    }
}

pub fn parse(text: &str) -> Tree {
    let tokens = tokenize_with_whitespace(text);
    let mut p = Parser::new(tokens);
    parse_sql(&mut p);
    p.build_tree()
}

// Parse a SQL file (entry point)
fn parse_sql(p: &mut Parser) {
    let m = p.open();

    while !p.eof() {
        if at_select_statement(p) {
            parse_select_statement(p);
        }

        if p.at(TokenKind::Semicolon) {
            p.expect(TokenKind::Semicolon);
        }
    }

    p.skip_trivia();

    p.close(m, TreeKind::File);
}

#[cfg(test)]
mod tests {
    use crate::parser::parser::parse;
    use rstest::rstest;

    #[rstest]
    fn test_parse(#[files("test/inputs/**/*.sql")] path: std::path::PathBuf) {
        let inputs_dir = std::path::Path::new("test/inputs").canonicalize().unwrap();
        let snapshots_dir = std::path::Path::new("test/snapshots")
            .canonicalize()
            .unwrap();

        let path_str = path
            .strip_prefix(inputs_dir)
            .unwrap()
            .to_str()
            .expect("Failed to convert path to string");
        let file_content =
            std::fs::read_to_string(&path).expect(&format!("Failed to read file: {}", path_str));

        // Try to parse the file content, catching any panics so that the panic message can be snapshotted
        let parse_result = std::panic::catch_unwind(|| parse(&file_content));

        // Compare the result with the snapshot
        insta::with_settings!({
            description => path_str,
            snapshot_suffix => "parse",
            snapshot_path => &snapshots_dir,
            prepend_module_to_snapshot => false,
            omit_expression => true,
        }, {
            match &parse_result {
                Ok(tree) => insta::assert_yaml_snapshot!(path_str, tree),
                Err(err) => insta::assert_yaml_snapshot!(path_str, err.downcast_ref::<&str>().unwrap().to_string()),
            }
        });
    }
}
