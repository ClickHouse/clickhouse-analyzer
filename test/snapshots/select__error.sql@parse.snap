---
source: src/parser/parser.rs
description: select/error.sql
---
parser is stuck
